Issue #5086 Review and rework o.e.j.util.Scanner

janbartel opened the pull request:
Closes #5086 

The issue task was review the locking strategy for the Scanner. I've done that, made changes to hopefully remove any necessity for locking, and also removed some methods that were deprecated in jetty-9.4.  As part of the refactoring of the code, there were some behaviour changes in the reporting of file changes, which I don't believe will make any substantive difference ( I can't even remember why I put those specific behaviours in there in the first place) but I'll call them out here anyway:

1. previously, we never reported a file as having been REMOVED until it had been missing for 2 scan cycles. I think that's unnecessary: when any scan detects a file has been deleted, a REMOVED event is issued for it.
1. previously, if a file was added and then deleted in the next scan then subsequently popped back into existence it was reported as CHANGED.  Now, if the file pops back into existence, it will reported as ADDED when it stabilizes, and REMOVED when it is deleted.

Note when reviewing: as a reminder of how the Scanner works, it is still the case that it takes at least 2 scan cycles to report a file as being ADDED: the first to detect the addition of the file, and the second to check that the file size or last modified time hasn't changed.  Similarly, it takes at lesat 2 scan cycles to report a file as CHANGED: the first to detect the change, and the second to check the file size or last modified time hasn't changed.

janbartel committed:
Issue #5086 Review and rework o.e.j.util.Scanner

Signed-off-by: Jan Bartel <janb@webtide.com>

janbartel committed:
Issue #5086 Improve ScannerTest

Signed-off-by: Jan Bartel <janb@webtide.com>

janbartel committed:
Issue #5086 Fix checkstyle in test

Signed-off-by: Jan Bartel <janb@webtide.com>

janbartel committed:
Issue #5086 Fix javadoc

Signed-off-by: Jan Bartel <janb@webtide.com>

sbordet commented on jetty-ant/src/main/java/org/eclipse/jetty/ant/ServerProxyImpl.java:
```diff
@@ -135,7 +136,7 @@ public WebAppScannerListener(AntWebAppContext awc)
 
         @Override
-        public void filesChanged(List<String> changedFileNames)
+        public void filesChanged(Set<String> changedFileNames)
```
Don't like this. File names are unique by definition, so there is no risk of duplicates. I would not add additional randomness caused by `Set`. If I really don't care, I'd use `Collection` -- the `Set` is supposed to be immutable anyway, right?

janbartel replied:
Not sure I understand your point. If it's a List<String> then by definition I can have duplicates in a List, but I can't in a Set. Plus, you can't rely on ordering as it depends on the filesystem in what order scanned directories are visited, so List is a furphy.

sbordet commented on jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ScanningAppProvider.java:
```diff
@@ -140,7 +140,6 @@ protected void doStart() throws Exception
         _scanner = new Scanner();
         _scanner.setScanDirs(files);
         _scanner.setScanInterval(_scanInterval);
-        _scanner.setRecursive(_recursive);
```
`Scanner.setRecursive()` was deprecated, but there is no Javadoc about why.
This is the only usage of `ScanningAppProvider._recursive` which then should also be deprecated and removed?

janbartel replied:
Done.

sbordet commented on jetty-util/src/test/java/org/eclipse/jetty/util/ScannerTest.java:
```diff
@@ -120,6 +122,18 @@ public Event(String filename, Notification notification)
+        public boolean equals(Object obj)
+        {
+            return ((Event)obj)._filename.equals(_filename) && ((Event)obj)._notification == _notification;
+        }
```
Add correspondent `hashCode()`. Use `Objects.hash(...)` if possible.

janbartel replied:
Done.

sbordet commented on jetty-util/src/test/java/org/eclipse/jetty/util/ScannerTest.java:
```diff
@@ -296,31 +309,31 @@ public void testAddedChangeRemove() throws Exception
+        assertThat(actualEvents, Matchers.containsInAnyOrder(a1, a3));
         assertTrue(_queue.isEmpty());
-
+        
```
Spaces!

janbartel replied:
Done.

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/Scanner.java:
```diff
@@ -67,27 +66,27 @@
-    private final List<Listener> _listeners = new ArrayList<>();
-    private final Map<String, TimeNSize> _prevScan = new HashMap<>();
-    private final Map<String, TimeNSize> _currentScan = new HashMap<>();
+    private AtomicInteger _scanCount = new AtomicInteger(0);
```
Make it `final`.

janbartel replied:
Done.

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/Scanner.java:
```diff
@@ -661,146 +609,116 @@ public boolean exists(String path)
      * Scan all of the given paths.
      */
-    public void scanFiles()
+    public Map<String, MetaData> scanFiles()
```
Make it `private` because `MetaData` is package private.
Maybe make `MetaData` a `private` class too?

janbartel replied:
Done.

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/Scanner.java:
```diff
@@ -271,9 +270,13 @@ public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOEx
 
+    /**
+     * Notification of files that changed in the last scan.
+     *
```
Remove extra newline.

janbartel replied:
Done.

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/Scanner.java:
```diff
@@ -312,125 +312,111 @@ public int getScanInterval()
-            _scanInterval = scanInterval;
-            schedule();
-        }
+    {        
```
Spaces.

janbartel replied:
Done.

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/Scanner.java:
```diff
@@ -544,30 +508,22 @@ public void removeListener(Listener listener)
+        schedule();
     }
 
     public TimerTask newTimerTask()
```
Make it `private` and replace with Jetty's `Scheduler`.

janbartel replied:
Done.

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/Scanner.java:
```diff
@@ -312,125 +312,111 @@ public int getScanInterval()
      */
     @Deprecated
-    public void setRecursive(boolean recursive)
+    public void setFilenameFilter(FilenameFilter filter)
```
Still worth to keep around `Deprecated` method?

janbartel replied:
I don't really want to mess with the WebAppProvider's filter for finding webapps to deploy at this late stage - it can be refactored at a later stage, so I've raised #5748 to track it.

janbartel committed:
Issue #5086 Changes after review

Signed-off-by: Jan Bartel <janb@webtide.com>

janbartel commented:
@sbordet I've done most if not all of things you requested so please rereview when ready. The failed build doesn't seem to have anything to do with this PR.

janbartel committed:
Issue #5086 Fix error on addFile

Signed-off-by: Jan Bartel <janb@webtide.com>

janbartel commented:
> @sbordet I've done most if not all of things you requested so please rereview when ready. The failed build doesn't seem to have anything to do with this PR.

OOps, looks like it was actually a real problem, just heavily disguised in build paraphernalia. Should be fixed now.

janbartel committed:
Merge remote-tracking branch 'origin/jetty-10.0.x' into jetty-10.0.x-5086-review-scanner

janbartel committed:
Issue #5086 Add blocking method for immediate scan.

Signed-off-by: Jan Bartel <janb@webtide.com>

sbordet committed:
Fixed scan(Callback) bug, returning early in case of failures.

More cleanups in code adding more privateness, getting rid of unnecessary exceptions, making fields final, etc.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet commented:
@janbartel I went on and polished this PR even more, you may want to review my commit.

gregw commented on jetty-util/src/main/java/org/eclipse/jetty/util/ssl/KeyStoreScanner.java:
```diff
@@ -135,8 +135,7 @@ public void scan()
             _scanner.scan(callback);
-            complete.await(10, TimeUnit.SECONDS);
-
+            return complete.await(10, TimeUnit.SECONDS);
```
@sbordet this will return true even if the callback fails, which  is a bit strange.

sbordet replied:
@gregw I don't understand? If the 10 s elapsed, `complete.await()` would have returned `false`, and the return value would have been ignored. Now we return it so that a JMX call to this method would know if it timed out.

gregw replied:
Scan is passed a callback. If it is failed then this method should return false.   Moreover I think if it times out then I think this method should throw a timeout exception rather than return

gregw replied:
@sbordet ^

Also I don't like the 10second wait either, but not sure how to avoid other than infinite wait

sbordet committed:
Using capital case for static final field.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Improved KeyStoreScanner.scan().

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>
